### PR TITLE
✨ amp-render refresh action

### PR DESCRIPTION
**Pull request by dmanek**
Partial for #32477

### CODE REVIEW

File: extensions/amp-render/1.0/amp-render.js
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:166]: Is `refresh` a forbidden term?

File: extensions/amp-render/1.0/test/test-amp-render.js
Comment (caroqliu) [inline extensions/amp-render/1.0/test/test-amp-render.js:189]: Suggestion: tests cases that refresh does nothing for the guarded conditions (`!src || isAmpStateSrc(src) || isAmpScriptSrc(src)`)

File: extensions/amp-render/1.0/amp-render.js
Comment (caroqliu) [inline extensions/amp-render/1.0/amp-render.js:164]: Suggestion: WDYT about storing `canRefresh` as an internal variable and updating in a `checkPropsPostMutations`? It doesn't seem necessary to calculate this every call unless `src` actually changes between refresh actions, but then again, maybe it is fine to repeat this logic. We also follow the same conditional branches in `getJsonFn`, would be nice if there was an even simpler way than what I suggested. 🤔

**Review by caroqliu — COMMENTED** [review]
LGTM, had a couple questions/suggestions and request for additional test cases.

File: extensions/amp-render/1.0/amp-render.js
Comment (krdwan) [reply extensions/amp-render/1.0/amp-render.js:164]: `mutationObserverCallback` may also be helpful here if you plan to change the logic.  This method is called each time `attributes` change.  You can keep an instance variable of `src` and `canRefresh` (or however you'd like to store state) and it'll be synced with the `src` attribute on the element.

Here is an example usage: https://github.com/ampproject/amphtml/blob/main/extensions/amp-social-share/1.0/amp-social-share.js#L157

Depending on your config, the callback is also called on changes to `childList` and potentially other options so you might want to double check you update state only when needed.

**Review by krdwan — COMMENTED** [review]
LGTM, pending Caroline's suggestions.

File: extensions/amp-render/1.0/amp-render.js
Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js:164]: So mimicking the behavior of `amp-list` here - we should allow refreshing even though `src` is unchanged (there is an amp-list example in the docs on amp.dev).

File: extensions/amp-render/1.0/amp-render.js
Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js:166]: copy-paste remnant.

File: extensions/amp-render/1.0/test/test-amp-render.js
Comment (dmanek) [reply extensions/amp-render/1.0/test/test-amp-render.js:190]: Done.

File: extensions/amp-render/1.0/amp-render.js
Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js]: I thought about this a bit more - isn't is better to bail here on every `refresh` action rather than re-compute `canRefresh` on every `mutationObserverCallback` which involves iterating thru all the mutation records?

Comment (caroqliu) [reply extensions/amp-render/1.0/amp-render.js]: You don't have to iterate through all mutation records, only read the final `getAttribute('src')` state:
```js
  mutationObserverCallback() {
    const open = this.element.getAttribute('src');
    if (src === this.src_) {
      return;
    }
    this.src_ = src;
    this.canRefresh_ = !src?.length || (...etc);
  }
```

Anyway, it is up to you and likely depends on how the author plans on using the component, i.e. if they never implement amp-render with `refresh` action but mutate `src` often, then it is more costly to compute in `mutationObserverCallback` and vice versa for placing the logic in the `registerApiAction` callback. The rest of the PR looks good to me so no need to block on this detail.

**Review by caroqliu — APPROVED** [review]

File: extensions/amp-render/1.0/amp-render.js
Comment (krdwan) [reply extensions/amp-render/1.0/amp-render.js]: I think either way is fine.  Just wanted to provide you another option.

I suppose it really comes down to which one is expected to change more or if you plan to use instance variable of `canRefresh` and `src` in other parts of the component.

**Review by krdwan — APPROVED** [review]
LGTM

File: extensions/amp-render/1.0/amp-render.js
Comment (dmanek) [reply extensions/amp-render/1.0/amp-render.js]: Appreciate the alternate solution here. I've added a comment for my future self linking to this thread, so I can revisit if necessary.

**Pull request merged**
